### PR TITLE
add parser

### PR DESCRIPTION
**Pull request by eduardpauliuc**

### CODE REVIEW

File: Parser/pom.xml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    <properties>
+        <maven.compiler.source>15</maven.compiler.source>
+        <maven.compiler.target>15</maven.compiler.target>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.20</version>
```
Comment (alex-nedelcu) [inline Parser/pom.xml:20]: Having the dependencies versions defined within the `<properties>` tag is the recommended way:
```
<properties>
  <lombok.version>1.18.20</lombok.version>
</properties>
```
and using them as `<version>${lombok.version}</version>` inside the `<dependency>` tag

File: Parser/src/main/java/Grammar.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+public class Grammar {
+
+    private NonTerminal initialSymbol;
+    private List<NonTerminal> nonTerminals;
+    private List<Terminal> terminals;
+    private List<Production> productions = new ArrayList<>();
+
+    private Pair<List<String>, List<String>> splitProduction(String production) {
+        List<String> splittedProduction = List.of(production.split("->"));
+        return new Pair<>(Arrays.stream(splittedProduction.get(0).strip().split(" ")).collect(Collectors.toList()),
```
Comment (alex-nedelcu) [inline Parser/src/main/java/Grammar.java:19]: Code readability could be improved here and in all the similar methods

File: Parser/src/main/java/Grammar.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    public boolean isNonTerminal(String label) {
+        return this.nonTerminals.stream().map(NonTerminal::getLabel).anyMatch(l -> Objects.equals(label, l));
+    }
+
+    private void loadProductions(List<String> productions) throws ReadGrammarException {
+        String production;
+        Pair<List<String>, List<String>> splittedProduction;
+        for (var i = 0; i < productions.size(); i++) {
+            production = productions.get(i);
+            splittedProduction = splitProduction(production);
```
Comment (alex-nedelcu) [inline Parser/src/main/java/Grammar.java:40]: In general it's good to have the methods used inside another method right below it, in the order in which they are used, such that the code can be read like a newspaper🤌🏼

Comment (alex-nedelcu) [reply Parser/src/main/java/Grammar.java:40]: e.g. move `splitProduction` below `loadProduction`

File: Parser/src/main/java/Parser.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+public class Parser {
+
+    private Grammar grammar;
+    private ParsingTable table;
+    Stack<Object> workingStack = new Stack<>();
```
Comment (alex-nedelcu) [inline Parser/src/main/java/Parser.java:22]: Maybe you forgot the access modifiers for the stacks and tree

File: Parser/src/main/java/Grammar.java (hunk 2 above)
Comment (calin2110) [reply Parser/src/main/java/Grammar.java:19]: I definitely with Alex here. I'd make `Arrays.stream(string).strip().split("")).collect(Collectors.toList())` a separate function, while also checking the length before to be exactly 2 to have no surprises

File: Parser/src/main/java/Grammar.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        }
+    }
+
+    private List<String> splitLine(String line) throws ReadGrammarException {
+        List<String> splitted = new ArrayList<>();
+        StringTokenizer tokenizer = new StringTokenizer(line, " ", false);
+        while (tokenizer.hasMoreTokens()) {
+            var token = tokenizer.nextToken().trim();
+            splitted.add(token.trim());
```
Comment (calin2110) [inline Parser/src/main/java/Grammar.java:56]: why do we trim the token twice?

File: Parser/src/main/java/Grammar.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        try {
+            fileScanner = new Scanner(file);
+            try {
+                while (fileScanner.hasNextLine()) {
+                    lines.add(fileScanner.nextLine());
+                }
+                this.terminals = splitLine(lines.get(0)).stream().map(Terminal::new).collect(Collectors.toList());
+                this.nonTerminals = splitLine(lines.get(1)).stream().map(NonTerminal::new).collect(Collectors.toList());
+                this.initialSymbol = new NonTerminal(splitLine(lines.get(2)).get(0));
+                System.out.println(lines.subList(3, lines.size()));
```
Comment (calin2110) [inline Parser/src/main/java/Grammar.java:77]: is this print call needed?

File: Parser/src/main/java/Grammar.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                loadProductions(lines.subList(3, lines.size()));
```
Comment (calin2110) [inline Parser/src/main/java/Grammar.java:78]: don't we want to check size while we're at it?

File: Parser/src/main/java/Grammar.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    public List<Production> getProductionsForNonTerminal(String nonTerminalValue) {
+        var searchValue = nonTerminals.stream()
+                .filter(x -> x.getLabel().equals(nonTerminalValue)).findFirst();
+        if (searchValue.isEmpty())
+            throw new RuntimeException("Provided value is not a valid non-terminal");
+
+        var nonTerminal = searchValue.get();
+
+        return productions.stream().filter(p -> p.getLeftForCFG().equals(nonTerminal)).collect(Collectors.toList());
```
Comment (calin2110) [inline Parser/src/main/java/Grammar.java:99]: you might want to split this on 2 or more lines :?

File: Parser/src/main/java/Item.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        var item = new Item();
+
+        item.setLeft(left);
+        var newWork = new ArrayList<>(this.work);
+        newWork.add(this.input.get(0));
+        item.setWork(newWork);
+
+        var newInput = new ArrayList<>(this.input);
+        newInput.remove(0);
+        item.setInput(newInput);
```
Comment (calin2110) [inline Parser/src/main/java/Item.java:54]: I would also check the size there

File: Parser/src/main/java/Main.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+public class Main {
+
+
+    public static void main(String[] args) {
+        var menu = new Menu();
+        menu.Run();
```
Comment (calin2110) [inline Parser/src/main/java/Main.java:6]: we want to call main() here right?

File: Parser/src/main/java/Menu.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+                case "6" -> {
+                    System.out.println("is CFG? " + grammar.isCFG());
+                }
+
+            }
+        }
+    }
+
+    public void Run() {
```
Comment (calin2110) [inline Parser/src/main/java/Menu.java:82]: and then delete this

File: Parser/src/main/java/Node.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+public class Node {
+    private Integer index;
+    private String info;
+    private Integer parent;
+    private Integer rightSibling;
+
+    @Override
+    public String toString() {
+        return  "{ " + index + " } " + " { " + info + " } " + " { parent:" + parent+ " } " + " { sibling:" + rightSibling + " } ";
```
Comment (calin2110) [inline Parser/src/main/java/Node.java:17]: String.format?

File: Parser/src/main/java/Parser.java [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+            for (var item : oldClosure) {
+                if (!item.getInput().isEmpty()) {
+                    var symbol = item.getInput().get(0);
+                    if (grammar.isNonTerminal(symbol.getLabel())) {
+                        var oldSize = closure.size();
+                        closure.addAll(grammar.getProductionsForNonTerminal(symbol.getLabel())
+                                .stream()
+                                .map(Item::new)
+                                .collect(Collectors.toSet()));
+                        changed = changed | oldSize != closure.size();
```
Comment (calin2110) [inline Parser/src/main/java/Parser.java:52]: am I dumb or is this supposed to be ||

File: Parser/src/main/java/Parser.java [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    public void generateCanonicalCollection() {
+        this.enhanceGrammar();
+
+        this.states.add(this.closure(Set.of(
+                new Item(grammar.getInitialSymbol(),
+                        List.of(),
+                        grammar.getProductionsForNonTerminal(
+                                grammar.getInitialSymbol().getLabel()).get(0).getRight()))));
```
Comment (calin2110) [inline Parser/src/main/java/Parser.java:79]: closed parantheses on new lines

File: Parser/src/main/java/Parser.java [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+
+                    changed = true;
+                    newStates.add(goToResult);
+                }
+            }
+            this.states = newStates;
+        }
+    }
+
+    private Integer findState(Set<Item> state) {
```
Comment (calin2110) [inline Parser/src/main/java/Parser.java:99]: we want non-nullable => primitive int

File: Parser/src/main/java/Parser.java [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+        var symbolsStack = new Stack<Symbol>();
+        symbolsStack.addAll(symbols);
+        while (!symbolsStack.isEmpty()) {
+            try {
+                var currentS = workingStack.pop();
+                if (currentS.equals(symbolsStack.peek().getLabel())) {
+                    symbolsStack.pop();
+                }
+            } catch (Exception e) {
+                System.out.println("e");
```
Comment (calin2110) [inline Parser/src/main/java/Parser.java:172]: you probably want to print the exception here
you should also never do catch (Exception ...)

File: Parser/src/main/java/Parser.java [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+        var state = 0;
+        Collections.reverse(sequence);
+        inputStack.addAll(sequence);
+        workingStack.push(state);
+        var done = false;
+        do {
+            try {
+                state = (int) workingStack.peek();
+            } catch (Exception e) {
+                System.out.println("e");
```
Comment (calin2110) [inline Parser/src/main/java/Parser.java:242]: same as above

File: Parser/src/main/java/Parser.java [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+                    actionReduce();
+                } else {
+                    if (Objects.equals(table.getRows().get(state).getAction(), "accept")) {
+                        System.out.println("success!");
+                        var output = new ArrayList<>(outputStack);
+                        Collections.reverse(output);
+                        System.out.println(output);
+                        done = true;
+                    } else {
+                        System.out.println("erorr!");
```
Comment (calin2110) [inline Parser/src/main/java/Parser.java:257]: error

File: Parser/src/main/java/ParserOutput.java [hunk 19]
```diff
@@ -0,0 +1,2 @@
+public class ParserOutput {
```
Comment (calin2110) [inline Parser/src/main/java/ParserOutput.java]: why does this exist

File: Parser/src/main/java/ParsingTableRow.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    public void setGoToSymbol(Symbol symbol, Integer state){
+        for (var pair: goTo) {
+            if(pair.getFirst().equals(symbol)){
+                pair.setSecond(state);
+            }
+        }
+    }
+
+    public Integer getGoToState(String symbol){
+        for (var s: goTo) {
```
Comment (calin2110) [inline Parser/src/main/java/ParsingTableRow.java:34]: I am not sure what s is.

Comment (calin2110) [reply Parser/src/main/java/ParsingTableRow.java:34]: also, above when insuccess when finding state, we return -1, here we return null => inconsistency

**Review by calin2110 — COMMENTED** [review]
nothing personal

**Pull request merged**